### PR TITLE
1A - Map H parameters through in an algorithmic rather than hand crafted way

### PR DESCRIPTION
**Pull request by jon-betts**
Instead of manually mapping params through we now expect them to have a name which corresponds to the location they will end-up in in h's config.

For example:

   `via.h.some_option.subValue=foo`

Will become:

    {
        'some_option': {
            'subValue': 'foo
        }
    }

We no longer attempt to cast values to the correct types. This will require a change in the client to handle that for us. We also don't attempt to interpret the case in anyway, we pass what we get.

----

Requires: https://github.com/hypothesis/client/issues/1957 as we have stopped casting values

### CODE REVIEW

File: via/configuration.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        cls._move_legacy_params(via_params, h_params)
+
+        # Set some defaults
+        h_params["appType"] = "via"
+        h_params.setdefault("showHighlights", True)
+
+        return via_params, h_params
+
+    @staticmethod
+    def _unflatten(params):
```
Comment (LMS007) [inline via/configuration.py:72]: This function feels a bit specialize since it only breaks it down if the prefix is "via."

I'm wondering if this shouldn't just take a param called `prefix` or it should be called `_unflattenViaParams`

File: via/configuration.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    @classmethod
+    def _filter_h_params(cls, h_params):
+        """Remove keys which are not in the whitelist."""
+
+        for key in set(h_params.keys()) - cls.H_CONFIG_WHITELIST:
+            h_params.pop(key)
+
+    @staticmethod
+    def _move_legacy_params(via_params, h_params):
```
Comment (LMS007) [inline via/configuration.py]: Ya, gota get the client converted over , then we can do-away with this. But I think this is a good step.

File: via/configuration.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        # This is like to be around for a while
+        h_params.setdefault("openSidebar", via_params.pop("open_sidebar", False))
+
+        # This should be removed when LMS is refactored to send things the
+        # new way
+        request_config = via_params.pop("request_config_from_frame", None)
+        if request_config is not None:
+            h_params["requestConfigFromFrame"] = {
+                "origin": request_config,
+                "ancestorLevel": 2,
```
Comment (LMS007) [inline via/configuration.py]: This value should already be in the via url correct? 
`via.config_frame_ancestor_level`
We shouldn't need to re-create it here , instead can we just pass it through?

File: tests/unit/via/configuration_test.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import pytest
+from h_matchers import Any
+
+from via.configuration import Configuration
+
+
+class TestConfiguration:
```
Comment (LMS007) [inline tests/unit/via/configuration_test.py:7]: I'd like to see a 3-level nested test.

File: via/configuration.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            parts = key.split(".")
+            if parts[0] != "via":
+                continue
+
+            target = data
+            # Skip the first 'via' part and then recurse to the one before last
+            for part in parts[1:-1]:
+                target = target.setdefault(part, {})
+
+            # Finally set the last key to the value
```
Comment (LMS007) [inline via/configuration.py:92]: Can you add a note that value is always a string. Actually is it?

**Review by LMS007 — CHANGES_REQUESTED** [review]
Overall, I think this is moving in the right direction and a huge improvement. The only big question mark for me is rules around casting. I see that we're setting `showHighlights` to a bool, but then any "true" or "false" string in a nested structure will just be a string value ( I assume)

I'd like some agreement about types that relates to the final output js object and then I could like to see that in the docs for the `Configuration ` class. If everything is just always a string, okay, we can deal with that and backfill the client after this, but if some things are only string but others  are bool or numbers, then thats a bit confusing at first glance. 

As an alternative, we could had a second & optional delimiter that casts the value.
e.g.
`via.config_frame_ancestor_level.2:number`
`via.open_sidebar.true:bool`

Just as an example, but that also makes it a bit more complex. Thoughts?

File: via/configuration.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+        data = {}
+
+        for key, value in params.items():
+            parts = key.split(".")
+            if parts[0] != "via":
+                continue
+
+            target = data
+            # Skip the first 'via' part and then recurse to the one before last
```
Comment (LMS007) [inline via/configuration.py]: This is not recursion right? maybe change verb to "iterate"

File: tests/unit/via/configuration_test.py (hunk 4 above)
Comment (jon-betts) [reply tests/unit/via/configuration_test.py:7]: `test_it_extracts_nested_h_elements` has a 4 part string which results in a 3 part value which is copied from `h` to root resulting in a 2 part value. We could start with 5?

So internally this triggers the code to deal with a 3 level nested param.

File: via/configuration.py (hunk 1 above)
Comment (jon-betts) [reply via/configuration.py:72]: It's specialised, but it's also private to this class, so it only needs to differentiate itself amongst other functions here, rather than globally.

This function actually returns both `h` and `via` configuration, so I don't think it's accurate to put "Via" in the name. In general, as we _are_ via, it can be broadly assumed that without stating we're in a "Via" context for most things. So this object is "Configuration" rather than "ViaConfiguration" as it's obvious from context.

It might be good to mention in the config string that this only retrieves parameters prefixed with "via." though. The doc string reads as if it's general.

File: via/configuration.py (hunk 6 above)
Comment (jon-betts) [reply via/configuration.py]: We are iterating over the keys, but recursively descending into the object by repeatedly overwriting `target`. But you're right here, the comment is about the key iteration.

File: via/configuration.py (hunk 5 above)
Comment (jon-betts) [reply via/configuration.py:92]: It might happen to be, but in this function doesn't know or rely on that and it works equally well with anything.

File: via/configuration.py (hunk 2 above)
Comment (jon-betts) [reply via/configuration.py]: That's the plan! https://github.com/hypothesis/client/issues/1958

We'll probably still have the "openSidebar" part for the foreseeable, as 3rd parties are using this.

File: via/configuration.py (hunk 3 above)
Comment (jon-betts) [reply via/configuration.py]: That's kind of the idea of this PR. The value will not be `via.config_frame_ancestor_level` any more. It will be `via.request_config_from_frame.ancestor_level`. This will then be automatically translated into the right place by the combination of `_unflatten` and `_to_camel_case`.

So if you set it, this block won't trigger.

It also won't trigger for the moment if you pass just `via.request_config_from_frame` by itself. This is for backwards comparability while we execute this change. When this is done we can remove this step and rely on the client always passing through:

 * `via.request_config_from_frame.origin`
 * `via.request_config_from_frame.ancestor_level`

The full steps are in this ticket: https://github.com/hypothesis/client/issues/1958

File: via/configuration.py (hunk 6 above)
Comment (jon-betts) [reply via/configuration.py]: I've updated the comment

File: via/configuration.py (hunk 1 above)
Comment (jon-betts) [reply via/configuration.py:72]: I've updated the doc string to give details of the 'via' key skipping behavior.

**Comment by jon-betts** [comment]
I think there might be some missing context which is making this more difficult to review.

There are two main ideas here:

### Have a direct automated mapping from the param name and the final config location

  * This will require the parameters LMS is sending to change
  * We will not be sending `via.config_frame_ancestor_level` any more
  * There's no ticket for that yet, but the general plan is here: https://github.com/hypothesis/client/issues/1958

### Prevent us from having to update Via and Via 3 for every new parameter we add 

 * If we have to cast values to different types with different defaults in Via / Via 3, this won't happen
 * The plan is therefore for the client to cast and apply defaults for a few reasons
 * It's the clients parameters, so it knows what type and default should be applied (why should Via choose the clients defaults?)
 * The client should really be applying it's own cleaning when reading these values anyway. If it's a problem to read a string where an int should be, the client should enforce/coerce the type
 * The long and short of this is, via will always receive strings, and will always pass on strings, it will be up to the client to interpret them
 * Currently this only means two fields as detailed here: https://github.com/hypothesis/client/issues/1957

**Comment by jon-betts** [comment]
Oops. Wrong button

**Comment by jon-betts** [comment]
I've made some changes based on feedback from @seanh and @LMS007 :

 * It's not even H config, it's client config so that's been renamed
 * More doc strings
 * I've removed the case mapping. If we want it to be "blahBlahBlah" we can just put that in the param name too. This also means weird edge cases and exceptions don't cause us any bother

File: tests/unit/via/configuration_test.py (hunk 4 above)
Comment (LMS007) [reply tests/unit/via/configuration_test.py:7]: awe, missed that, never mind!

**Comment by LMS007** [comment]
@jon-betts , I think we're ready to move forward with this, was there anything blocking this anymore? Client should be ready now.

File: via/configuration.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    @classmethod
+    def _filter_client_params(cls, client_params):
+        """Remove keys which are not in the whitelist."""
+
+        for key in set(client_params.keys()) - cls.CLIENT_CONFIG_WHITELIST:
+            client_params.pop(key)
+
+    @staticmethod
+    def _move_legacy_params(via_params, client_params):
```
Comment (LMS007) [inline via/configuration.py:105]: We can probably do away with the type conversion here as this is now done on the client. 
e.g. this should work
```
"requestConfigFromFrame" = {
                "origin": request_config,
                "ancestorLevel": "2",
}
```

File: via/configuration.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        """Handle legacy params which we can't move for now."""
+
+        # This is like to be around for a while
+        client_params.setdefault("openSidebar", via_params.pop("open_sidebar", False))
```
Comment (LMS007) [inline via/configuration.py:109]: No need for this anymore

**Review by LMS007 — APPROVED** [review]
I think we can get rid of the type casting, then get this merged.

File: via/configuration.py (hunk 8 above)
Comment (jon-betts) [reply via/configuration.py:109]: I think we need both of these for now. These are the legacy handling for when LMS sends us the old values (as it does right now). The False here is our default from Via3, rather than a type coercion.

Comment (LMS007) [reply via/configuration.py:109]: Roger that

File: via/configuration.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+        # This should be removed when LMS is refactored to send things the
```
Comment (LMS007) [inline via/configuration.py:111]: do you mind adding a searchable string for all the places we need to remove code for the "new way"
It could even just be "new way" if that works? But it could also be helpful to prefix this with "FIXME:"

**Review by LMS007 — APPROVED** [review]
Just left 1 new comment, otherwise this looks good to merge and start the process of getting params converted over. Thanks Jon

File: via/configuration.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        # new way
+        request_config = via_params.pop("request_config_from_frame", None)
```
Comment (LMS007) [inline via/configuration.py:113]: do you mind adding a searchable string for all the places we need to remove code for the "new way"
It could even just be "new way" if that works? But it could also be helpful to prefix this with "FIXME:"

Comment (jon-betts) [reply via/configuration.py:113]: I've already removed the code in future PRs, so I'd prefer not to add anything more, as the code should be gone by the end of today.

This is pretty much the only place as well.
